### ISSUE TITLE
Update LDAP backend

### ISSUE THREAD

**Issue by Habbie** [PowerDNS/pdns#3358]
Commits in @habbie's LDAP branch, as a checklist, checked when part of a PR:
 * [ ] 61f102fe25013e4a614ce73fa152b64d17ea6693 get root dots
 * [ ] 19f0fefeb378321a29fc46d09dd64faf9376638b fix escaping
 * [ ] 6f16614c0ba750408d66c0dbfc89d4aedc907006 71 out of 84 (84.52%) tests passed, 74 were skipped
 * [ ] 5f917136eb06249478df89211a36155efd6aeecb allow testing of both tree and simple
 * [ ] 9d84a4d439990534eced5b04aac296e5ea1561f9 Write INSTALL doc
 * [ ] df79334f85f2d4dd99813714edda63f28adb5799 Refactor LDAP backend
 * [ ] 99a1ee71a835128f8b89d16bff10ba5ee89d536d Add LDAP simple authenticator
 * [ ] 1bcfa0e3ff692e4fc200b99c84e458ca7216b2b6 Add LDAP GSSAPI authenticator
 * [ ] f352b012407d60dae461fe124491a10747a4b34c Prefetch results in PowerLDAP::search()
 * [ ] 484c6a52ea9a4190c998e5b929a883c33b28e753 Don't throw an exception in ldapWaitResult()
 * [ ] 8f2b0e14094a0d192ab5869007887e8f7f4a26a6 New LDAP exception class
 * [ ] 819464060f365b85c99e4b68ced7255fe07e5d07 Try to reconnect to LDAP when connection is lost
 * [ ] 6f251cee4f777efffc39a352a739a8d969ff29a6 ChangeLog update
 * [ ] f8593069da9ffbf960d1c7a8b0f3189f7535a1fa Write README
 * [ ] a39e80c50223db266ecefb2840b71e935eb371ee Add USAGE file to document options
 * [ ] 83e0c67b6994016d68c5be72defd02d04b9cd990 Add dnsdomain2 schema
 * [ ] 888eda34e8dce45887841f3dbc9d1488ca70abb0 Update ChangeLog for v0.1
 * [ ] c974796bf3bdcc544a642c115e54ad3abc2307ce Export KRB5CCNAME from the ldap backend This makes unnecessary the modification to the startup script.
 * [ ] 773510c751d13415c659ac92be7a4143f3d36ce5 Add schema for DomainInfo objects
 * [ ] 3fcabe89043dfb7fab87103ad5add4f7234e32a2 Add master support
 * [ ] c98b4cf609711c5e92a2f2f569323b636aff7897 Add USAGE.master
 * [ ] 3a173edea20162544060cb4f420bed098e4e8ed3 Throw LDAPNoConnection() if needed here too
 * [ ] 9416fe7e684e5ba43e7f534c13d110f511091781 Prevent potential side effects and use local variables here
 * [ ] 4dae82321906da891ba5812bf8cd1dc54d6bda51 getDomainInfo() is public
 * [ ] cd654704e89e5de1220c920a2a1ef241c13931f7 Remove AC_PREREQ
 * [ ] 747245b11eb873e19520a1fbe127f623b9256d61 Add instruction about the use of --libdir
 * [ ] e45c46c5a0a8577bcbbb860a06e85b71262305eb s/Ahu/PDNS/
 * [ ] 942d0908c48f4098d9fb208ad9d7f62f168bf898 avoid use of lexical_cast
 * [ ] 57abb50d8320c8d9b5ace121ea137e745c0b09f1 Add a modify method to PowerLDAP
 * [ ] cc85a1e6679ff634624307c7cdc8227e7b3717b6 pull in Makefile changes
 * [ ] 70b11e721c47ca019dfe623c3b68d72034a20e3f osx compile hack REVERT ME
 * [ ] 3e4d09a2e197b9a1af0101dd3ff43e092ab6bf88 import schema files from the ldap fork
 * [ ] 07212cdac1ef8b57765ec85afa3eb33a45309019 add all missing types (as far as our tests are concerned) to schema and header
 * [ ] cd08d9faacd9e0085b48f3ee5d5f9e3cef1a756d reduce ldapmodify verbosity

Open issues:
 * [ ] TTL is per name instead of per RRset in the current schema(s)

**Comment by goestreicher** [comment]
Hey Peter,

So I've started porting all these patches on the PowerDNS master branch. I've forked the pdns repo (https://github.com/KamagoNet/pdns/) and created a topic branch named 'ldap-backend-update' to prepare a PR. Actually I'm early in this, so the only tests are "does it compile?", and I'll test with a full setup after.

If you have any feedback just shout!

Cheers,
Grégory

**Comment by goestreicher** [comment]
Hey,

So it's done for the port, with the exception of 3bcf9ab. What is the reason behind this? I don't have a Mac to test unfortunately.

Cheers,
Grégory

**Comment by Habbie** [comment]
Hey Gregory, thanks for the work! Can you open the PR so we don't lose track of it? I'll look at the Mac thing asap but likely not this week.

**Comment by goestreicher** [comment]
I'll just do some testing in a real environment to validate that it's still working fine before opening the PR if that suits you. At the moment everything compiles, so there's that. Hopefully I'll be done today.

**Comment by Habbie** [comment]
Sure thing! Note that if you open the PR (or enable travis-ci.org on your fork) the actual tests will be run for you.

**Comment by goestreicher** [comment]
Yup, I'd just like to test the Kerberos additions, and I have absolutely no idea how to setup a realm through Travis, while I have all my Ansible playbooks ready. Travis is the next step after I'm happy with the local testing.

**Comment by goestreicher** [comment]
Travis and I are equally happy with our tests, PR #4477 is available.

**Comment by goestreicher** [comment]
So addressing the issue "TTL is per name instead of per RRset in the current schema(s)" would allow the following tests to pass:
- any-query
- basic-ns-resolution
- mx-case-sensitivy-with-ap
- mx-with-simple-additional-processing

I will not fix this against the current master, but in the DNSSEC topic branch that will get its own PR. Doing it this way is, at this stage, easier for me.

I'm adding this info here so that it doesn't get lost, and creating a new issue doesn't seem worth it, though if you like I can open one.

**Comment by Habbie** [comment]
@goestreicher can you tell me what commits to check in the list at the top? Or perhaps, what commits not to check?

**Comment by goestreicher** [comment]
Here are the ones that you can check because they're in the PR:
4fc09b8
952c027
43d273b
2c9b568
2524bba
95238d1
19cee7e
5467fb6
4fc6ef1

These ones could be removed from the list as they were not taken from the fork or were squashed with other commits:
0fafcc4 (removed)
7aee3bc (squashed)
326d8ef (squashed)
74192db (squashed)
822f113 (squashed I think)
